File: 2019/02/2.py
import csv
cells = []
with open('input.csv', 'rt') as csvfile:
    r = csv.reader(csvfile, delimiter=',')
    for row in r:
        for cell in row:
            cells.append(int(cell))

def run(array):
    index = 0
    while array[index] != 99:

        if array[index+1] >= len(array) or array[index+2] >= len(array) or array[index+3] >= len(array):
            return [-1]

        if array[index] == 1:
            array[array[index+3]] = array[array[index+1]] + array[array[index+2]]
        elif array[index] == 2:
            array[array[index+3]] = array[array[index+1]] * array[array[index+2]]
        index += 4
    return array

part1 = cells.copy()
part1[1] = 12
part1[2] = 2

answer = 0
noun = -1
verb = 0

while answer != 19690720:
    if noun < 99:
        noun += 1
    else:
        noun = 0
        verb += 1

    trial = cells.copy()
    trial[1] = noun
    trial[2] = verb
    answer = run(trial)[0]

print(run(cells)[0], (100*noun)+verb)

File: 2019/02/2.jl
using DelimitedFiles

const raw_program = readdlm("input.csv", ',', Int)
const program = reshape(raw_program, length(raw_program))
const TARGET = 19690720

function p1(p, noun, verb)
	ip = 0
	p[2] = noun
	p[3] = verb
	while true
		op = p[ip+1]
		
		if op == 1
			p[p[ip+4]+1] = p[p[ip+2]+1] + p[p[ip+3]+1]
		elseif op == 2
			p[p[ip+4]+1] = p[p[ip+2]+1] * p[p[ip+3]+1]
		elseif op == 99
			break
		end
		ip += 4
	end
	return p[1]
end

function p2(p)
	for i in 0:99
		for j in 0:99
			if p1(copy(p), i, j) == TARGET
				return 100*i + j
			end
		end
	end
end

println(p1(copy(program), 12, 2))
println(p2(program))
